design/token-exchange-spi.md

```diff
@@ -0,0 +1,86 @@
+# Token Exchange SPI
+
+## Abstract
+This document outlines the design of the proposed Token Exchange SPI for Keycloak (TODO: JIRA link).
+## Rationale
+Keycloak does implement the OAuth 2.0 Token Exchange ([RFC 8693](https://tools.ietf.org/html/rfc8693)), but does that in a peculiar way (Securing Applications and Services Guide, [7. Token Exchange](https://www.keycloak.org/docs/latest/securing_apps/index.html#_token-exchange)):
+> Token exchange in Keycloak is a very loose implementation of the OAuth Token Exchange specification at the IETF. We have extended it a little, ignored some of it, and loosely interpreted other parts of the specification.
+
+Current implementation has the following limitations and shortcomings:
+* The following (optional) parameters are not supported:
+	* `resource`
+	* `scope`
+	* `actor_token` and `actor_token_type`
+* Supporting internal exchange to/from non-OAuth clients requires core Keycloak modification. Currently, only exchange to SAML is supported;
+* Exchanging to/from external IdPs is supported for OAuth IdPs only and requires brokering to be set up;
+* Token exchange is built into `TokenEndpoint` which is already ~1400 LoC.
+
+Considering the above, we are suggesting the introduction of the Token Exchange SPI.
+
+## Use Cases and Benefits
```

Comment:
I think we should group use cases accordingly to the different token exchange scenarios they are associated with:

* **Impersonation**: Not sure if covered in this section but one of the use cases we support. This one is probably what resource servers are mainly interested in in order to call other services assuming the role of the original client to which the token was issued for. The current implementation needs to be checked whether it is respecting the semantics for this scenario when issuing tokens. For instance, the `client_id` claim.
* **Delegation**: This is the scenario that I think is the most problematic in terms of compliance with specs. Even in docs we talk about delegation in terms of impersonation, where, for RFC 8693, delegation is related to the subject delegating to an actor to act on his behalf. At this regard, our current implementation lacks a lot of the semantics defined by the spec.
* **Cross-domain**: This is the scenario where tokens are exchanged to/from different domains. Currently, we have this trust scenario solely based on brokering, where you can exchange an internal to external, and external to internal tokens.

Your use cases seem to be related to the **Trust** scenario where:

* Complement Protocol SPI: non-OAuth/OIDC clients are by nature outside the boundaries of a realm.
* Complement Identity Provider SPI: non-Auth Idps are outside the boundaries of a realm (similar to brokers)
* Non-brokered trust: Probably the use case closer to the **Trust** scenario I'm mentioned. However, I'm not sure about the requirements around recognizing and validating tokens but exchanging.

For me, looks like we need to consider all those scenarios prior to discussing a re-design and new SPIs. All the use cases you defined seem to be related to trust. And I think the other scenarios might (or not) impact any decision we make about the design.

Reply:
I'd agree that our first use case is about cross-domain trust, but with elements of delegation thrown in for good measure. 
However, the idea behind the SPI is to go beyond this and make token exchange as flexible as possible, so that tokens can be exchanged for a variety of purposes. For example a token issued by a (non-standard) legacy system could be exchanged for a Keycloak token that allows access to a more modern API or system; or an access token can be exchanged at the system edge for an internal format understood by the protected API - so long as there is a Token Exchange SPI Provider loaded that understands the intent based on the provided parameters (audience, token type etc), then a wide variety of use cases should be possible.

Reply:
Looks like what we need then is not really a Token Exchange SPI but some SPI specific for token exchange that allows supporting different token types.

Based on your comment, I think we could:

* Implement a `TokenProvider`, where the implementation is targeted for parsing a specific token type (e.g.: `subject_token_type` == `urn:mytoken`)
* Parse, validate and extract information from the `subject_token`. Here also lies the place where a provider can be used for cross-domain trust.
* Create a generic representation for the incoming token
* Elect a `TokenProvider` based on the `requested_token_type` (or maybe another interface specific for this flow)
* Invoke methods on the elected implementation to transform/exchange the representation into the requested token type

Considering the current implementation, the code we have today to generate the token response for the token types we support would be within a specific token provider implementation (and not hardcoded into the token endpoint). The same goes for the code parsing the `subject_token`.

Note, however, that token exchange is flexible in regards to the response.

```
access_token
      REQUIRED.  The security token issued by the authorization server
      in response to the token exchange request.  The "access_token"
      parameter from Section 5.1 of [RFC6749] is used here to carry the
      requested token, which allows this token exchange protocol to use
      the existing OAuth 2.0 request and response constructs defined for
      the token endpoint.  The identifier "access_token" is used for
      historical reasons and the issued token need not be an OAuth
      access token.
```

So we can include whatever token we want into the `access_token` response claim.

I think something like that makes more sense than a Token Exchange SPI.

Reply:
```
interface TokenExchangeProviderFactory {
    
   /**
   * Returns the token types that can be handled by this provider.
   */
    String[] getSupportedTokenTypes();

   /**
   * Returns the token types that can be issued by this provider.
   */
    String[] getSupportedRequestedTokenTypes();

    /**
     * Returns the actor token types supported by this provider for delegation scenarios.
     */
    String[] getSupportedActorTokenTypes();
}

interface TokenExchangeProvider {

     /**
      * Validates the token request and returns a {@link TokenExchangeContext} holding the information
      * from the token that should be exchanged.
      */
     TokenExchangeContext validate(TokenRequest) throws TokenExchangeException;

     /**
      * Perform the token exchange based on the information from the {@link TokenExchangeContext}.
      */
     TokenExchangeResponse exchange(TokenExchangeContext);
}

/**
* Represents the information associated with a token exchange request
*/
interface TokenExchangeRequest {
    // access to token exchange request params
}

/**
* The context represents a valid token request, providing additional information that can be used by other providers
* to exchange a subject token to a requested token type.
*
* It should be possible to have multiple sub-types of this interface representing a specific token exchange context. For instance, an OAuth2TokenExchangeContext representing an OAuth2 subject token that can be exchanged for another (or the same) token type.
*/
interface TokenExchangeContext {
   
    // information about the audience
    // information about the resource
    // information about the scope
    // information about the subject_token, which should include the subject, and the client asking for the exchange, perhaps the original issuer, etc.
    // information about the actor trying to act on behalf of the subject, if in a delegation scenario
}

interface TokenExchangeResponse {
   void setIssuedTokenType(String)
   void setIssuedToken(String);
   void setTokenType(String)
   void setExpiration(int);
   void setRefreshToken(String);
   void setScope(String);
}

This API should allow us to support token types depending on the use case. For instance, we would provide OOTB an `OAuth2TokenProvider` that is capable of validating and exchanging Keycloak tokens. If you want to support exchanging an internal token with an OAuth2 token, then you just need to implement the `validate` method in another provider implementation, leveraging the logic for generating/exchanging OAuth2 tokens to the OOTB provider.

The same goes if you want to exchange an OAuth2 token with an internal token. In this case, you just need a provider that implements the `exchange` method to build your internal token, leveraging the validation and representation from the default OAuth2 token provider.

Hope it makes sense. Like I said, quite similar to the API herein proposed. But with the possibility to re-use providers depending on the use case.